### PR TITLE
Output dicts support in text generation pipeline

### PR DESCRIPTION
**Pull request by jonasrohw**
# What does this PR do?

It is a minor fix to the text generation pipeline. When calling with the generation argument `return_dict_in_generate=True,` the code breaks because it does not expect a dict from the `model.generate(...)` call. If you want to view logits, for example, a dict is required. This fix can handle `return_dict_in_generate=True` as a pipeline parameter.

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

@Rocketknight1

### CODE REVIEW

**Review by Rocketknight1 — COMMENTED** [review]
Hi - firstly, thanks for the PR, but I notice that in the code there's a lot more than just adding output dict support! It seems like you're also changing the handling for batched generation a lot.

I can understand the rationale there, but I think it would be better to split that into a separate PR. In other words, make this into a smaller PR that just adds output dict support, with a second PR focused on batching. WDYT?

**Comment by jonasrohw** [comment]
@Rocketknight1 Hi. Thanks for looking at the PR and pointing that out. I agree with you; I got a bit ahead of myself. I've removed most of the changes and just made the changes to support dictionary output. I'd be happy if you could take a look. Thanks again!

**Review by Rocketknight1 — APPROVED** [review]
I think this looks good to me now, and sorry for the delay! @gante since you're probably more familiar with `return_dict_in_generate`, can you take a look and make sure it's all okay before I ping a core maintainer?

**Comment by jonasrohw** [comment]
@gante Friendly reminder if you could take a look. Thank you very much!

**Comment by Rocketknight1** [comment]
Hang on, I'll get him!

**Comment by jonasrohw** [comment]
@Rocketknight1 Any success in finding him?

**Comment by gante** [comment]
@jonasrohw I exist, but I'm being pinged in many places :D

**Review by gante — APPROVED** [review]
LGTM, thank you for expanding the capabilities of the pipeline :D

**Comment by gante** [comment]
(fixed conflicts, will merge as soon as CI gets green)
